mmf/trainers/base_trainer.py

```diff
@@ -278,16 +281,38 @@ def _forward_pass(self, batch):
         return report
 
     def _backward(self, loss):
-        self.optimizer.zero_grad()
+        # for update_frequency > 1, scale the loss so that accumulated gradients are the expected magnitude
+        loss = loss / self.config.training.update_frequency
         loss.backward()
+        self.profile("Backward time")
+
+    def _check_start_update(self, epoch_batch_index, epoch_num_batches):
+        # When epoch_num_batches is not a multiple of the effective batch size, the last few batches
+        # will be discarded. I.e., they won't contribute to an update.
+        epoch_num_updates = epoch_num_batches // self.config.training.update_frequency
+        is_batch_discarded = (epoch_batch_index >= epoch_num_updates * self.config.training.update_frequency)
+        if is_batch_discarded:
+            return
 
-        if self.should_clip_gradients:
-            clip_gradients(self.model, self.num_updates, self.tb_writer, self.config)
+        is_first_batch_for_update = epoch_batch_index % self.config.training.update_frequency == 0
+        if is_first_batch_for_update:
+            self.current_iteration += 1
+            self.writer.write(self.num_updates + 1, "debug")
+            self.optimizer.zero_grad()
 
-        self.optimizer.step()
-        self._run_scheduler()
-        self.num_updates += 1
-        self.profile("Backward time")
+    def _check_finish_update(self, epoch_batch_index, report):
+        should_break = False
+        is_last_batch_for_update = (epoch_batch_index + 1) % self.config.training.update_frequency == 0
+        if is_last_batch_for_update:
+            if self.should_clip_gradients:
+                clip_gradients(self.model, self.num_updates, self.tb_writer, self.config)
+            self.optimizer.step()
+            self._run_scheduler()
+            self.num_updates += 1
+            should_break = self._logistics(report)
```

Comment:
Note this report (including the loss) is only from a single recent batch; it doesn't represent the entire update.

Reply:
Note also that this loss in this report is *not* the scaled-down loss (I mean the result of the division on L285).

Reply:
Can we explicitly update the loss to the scaled down loss by assigning it directly? Wouldn't that fix the issue?

Reply:
Can you elaborate? Are you saying you want to save the scaled-down loss in the report?